OpenSim/Simulation/StatesTrajectory.h

```diff
@@ -0,0 +1,641 @@
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#include <vector>
+#include <Simbody.h>
+
+#include <OpenSim/Common/Exception.h>
+
+#include "osimSimulationDLL.h"
+
+namespace OpenSim {
+
+/** Helper class to use range for loops with a pair of iterators. This class
+ * should only be used when you're sure the iterators are valid. Don't use this
+ * class directly; instead, use makeIteratorRange(). */
+// http://stackoverflow.com/questions/6167598/why-was-pair-range-access-removed-from-c11
+template <class Iterator>
+class IteratorRange {
+public:
+    IteratorRange(Iterator first, Iterator last)
+        : m_first(first), m_last(last) {}
+    Iterator begin() const { return m_first; }
+    Iterator end() const { return m_last; }
+private:
+    const Iterator m_first;
+    const Iterator m_last;
+};
+
+/** Make an IteratorRange object to be used in a range for loop
+ * @code
+ * for (auto& x : makeIteratorRange(v.begin(), v.end())) {
+ *     ...
+ * }
+ * @endcode
+ * @relates IteratorRange
+ */
+template <class Iterator>
+IteratorRange<Iterator> makeIteratorRange(Iterator first, Iterator last) {
+    return IteratorRange<Iterator>(first, last);
+}
+
+
+class Storage;
+class Model;
+
+
+// Design note: This class is part of OpenSim instead of Simbody since Simbody
+// users are likely to be interested in a more general State container that
+// doesn't assume the states are sequential in time.
+
+/** A sequence of SimTK::State%s that can be saved to a plain-text OSTATES
+ * file.  The trajectory can also be populated from such a file. You may obtain
+ * a StatesTrajectory from a simulation, or other numerical methods whose task
+ * is to produce a trajectory of states. Users can modify the trajectory by
+ * appending states to it, but users cannot modify the individual states that
+ * are already in the trajectory.
+ *
+ * This class was introduced in OpenSim version 4.0, and enables scripting
+ * (Python/MATLAB) and C++ users to postprocess their results much more
+ * flexibly than with an Analysis.
+ *
+ * ### Guarantees
+ * This class is designed to ensure the following:
+ * - The states are ordered nondecreasing in time (adjacent states *can* have
+ *   the same time).
```

Comment:
I was on the fence about this, and I'm happy to change it. Any other opinions out there?

To some extent, allowing the states to have the same time allows one to use the trajectory as just any container of states (perhaps multiple iterations of inverse kinematics for a single pose). On the other hand, if someone is appending multiple states with the same time, perhaps it means they are doing something unintended (a bug)? It may also be possible that some algorithms/simulations produce two final states with the same time; I feel like I may have seen this in the past.